File: backend/src/main/java/com/eduardo/backend/services/CompanyLinkService.java
package com.eduardo.backend.services;

import com.eduardo.backend.dtos.CompanyLinkRequestDTO;
import com.eduardo.backend.dtos.CompanyLinkResponseDTO;

import java.util.List;

/**
 * Define as regras de negócio do vínculo
 * entre usuários e empresas.
 */
public interface CompanyLinkService {

    // Lista empresas disponíveis para solicitação de vínculo
    List<CompanyLinkResponseDTO> getAvailableCompanies();

    // Cria uma solicitação de vínculo
    CompanyLinkResponseDTO requestAccess(CompanyLinkRequestDTO dto);

    // CLIENT visualiza seus próprios vínculos
    List<CompanyLinkResponseDTO> getMyLinks();

    // OWNER visualiza solicitações pendentes
    List<CompanyLinkResponseDTO> getPendingRequestsForOwner();

    // OWNER aprova uma solicitação
    CompanyLinkResponseDTO approve(Long linkId);

    // OWNER lista usuários aprovados de uma empresa
    List<CompanyLinkResponseDTO> getUsersLinkedToCompany(Long companyId);
}

File: frontend/src/app/core/services/companyLink.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { CompanyLinkResponseDTO } from '../models/company-link.model';
import { CompanyLinkRequirementsDTO } from '../models/company-link-requirements.model';
export type CompanyLinkRole = 'CLIENT' | 'DRIVER';

@Injectable({ providedIn: 'root' })
export class CompanyLinkService {
  private apiUrl = 'http://localhost:8080/api/company-links';

  constructor(private http: HttpClient) {}

  // CLIENT lista empresas aprovadas disponíveis para solicitar vínculo
  getAvailableCompanies(): Observable<CompanyLinkResponseDTO[]> {
    return this.http.get<CompanyLinkResponseDTO[]>(`${this.apiUrl}/available`);
  }

  // USER solicita vínculo com role (CLIENT ou DRIVER)
  requestAccess(
    companyId: number,
    requestedRole: CompanyLinkRole
  ): Observable<CompanyLinkResponseDTO> {
    return this.http.post<CompanyLinkResponseDTO>(
      `${this.apiUrl}/request`,
      {
        companyId,
        requestedRole,
      }
    );
  }

  // CLIENT vê seus vínculos
  getMyLinks(): Observable<CompanyLinkResponseDTO[]> {
    return this.http.get<CompanyLinkResponseDTO[]>(`${this.apiUrl}/mine`);
  }

  // OWNER vê solicitações pendentes
  getPendingRequests(): Observable<CompanyLinkResponseDTO[]> {
    return this.http.get<CompanyLinkResponseDTO[]>(`${this.apiUrl}/pending`);
  }

  // OWNER aprova solicitação
  approveRequest(linkId: number): Observable<CompanyLinkResponseDTO> {
    return this.http.put<CompanyLinkResponseDTO>(
      `${this.apiUrl}/${linkId}/approve`,
      null
    );
  }

  // OWNER vê usuários vinculados
  getUsersLinkedToCompany(
    companyId: number
  ): Observable<CompanyLinkResponseDTO[]> {
    return this.http.get<CompanyLinkResponseDTO[]>(
      `${this.apiUrl}/company/${companyId}/users`
    );
  }

  // Requisitos para criação de jornada
  getRequirements(): Observable<CompanyLinkRequirementsDTO> {
    return this.http.get<CompanyLinkRequirementsDTO>(`${this.apiUrl}/requirements`);
  }
}
